### PR TITLE
initial implementation of new ETI

### PR DESCRIPTION
**Pull request by jjwilke**

### CODE REVIEW

**Comment by jjwilke** [comment]
* Remove huge number of cpp files 
* Use CMake system to generate single source file per kernel
* Makefile system will no longer work

**Comment by jjwilke** [comment]
New PR instead of canceled #526

**Comment by srajama1** [comment]
@jjwilke : This is great ! Deleting 6000 files from the repo, what else can we ask for :) 

@crtrott : Can you review this, please ?

**Comment by ndellingwood** [comment]
Please make sure to test with Trilinos as well as the `cm_test_all_sandia` on various testbeds before considering a merge.

**Comment by jjwilke** [comment]
Testing to-dos:

-[ ] cm_test_all_sandia on kokkos-dev-2
-[ ] cm_test_all_sandia on blake
-[ ] cm_test_all_sandia on mayer
-[ ] Trilinos promotion test with kokkos/develop on waterman
-[ ] Trilinos promotion test with kokkos/develop on blake

File: CMakeLists.txt [hunk 1]
```diff
@@ -217,9 +217,9 @@ KOKKOSKERNELS_ADD_OPTION(
 )
 
 SET(EXECSPACE_CUDA_VALID_MEM_SPACES      CUDASPACE CUDAUVMSPACE)
-SET(EXECSPACE_SERIAL_VALID_MEM_SPACES    HBWSPACE HOSTSPACE CUDAUVMSPACE)
-SET(EXECSPACE_OPENMP_VALID_MEM_SPACES    HBWSPACE HOSTSPACE CUDAUVMSPACE)
-SET(EXECSPACE_PTHREAD_VALID_MEM_SPACES   HBWSPACE HOSTSPACE CUDAUVMSPACE)
+SET(EXECSPACE_SERIAL_VALID_MEM_SPACES    HBWSPACE HOSTSPACE)
```
Comment (ndellingwood) [inline CMakeLists.txt]: @jjwilke do these changes for valid mem spaces need to propagate to develop independent of this PR?

Comment (jjwilke) [reply CMakeLists.txt]: No, this is only an ETI thing.

**Comment by jjwilke** [comment]
I had to make quite a few changes for blake to various things, independent of the ETI.

**Comment by ndellingwood** [comment]
> I had to make quite a few changes for blake to various things, independent of the ETI.

@jjwilke can you cherry-pick these changes for a separate PR?

**Comment by jjwilke** [comment]
Cherry-picked changes here: https://github.com/kokkos/kokkos-kernels/pull/533

**Comment by jjwilke** [comment]
Timing data:
* Old ETI w/TESTS
  -real	2m38.381s
  -user	42m4.556s
  -sys	8m11.060s

File: CMakeLists.txt (hunk 1 above)
Comment (mhoemmen) [reply CMakeLists.txt]: so we're double-instantiating all those kernels?  sadness :(

Comment (jjwilke) [reply CMakeLists.txt]: I think we were before, too? This PR is a "preserve all existing behavior". The next PR would do the anon spaces stuff.

One thing at a time!

Comment (mhoemmen) [reply CMakeLists.txt]: @jjwilke praise Turing and pass the ammunition ;-)

**Comment by ndellingwood** [comment]
Unsuccessful test on kokkos-dev-2, spot-check with cm_test_all_sandia script

```
#######################################################
PASSED TESTS
#######################################################
#######################################################
FAILED TESTS
#######################################################
clang-8.0-Cuda_OpenMP-release (build failed)
clang-8.0-Pthread_Serial-release (build failed)
cuda-10.1-Cuda_OpenMP-release (build failed)
gcc-7.3.0-OpenMP-release (build failed)
gcc-7.3.0-Pthread-release (build failed)
gcc-9.1-OpenMP-release (build failed)
gcc-9.1-Serial-release (build failed)
intel-18.0.5-OpenMP-release (build failed)
```

Looks like linking errors with the trsm update:

```
Testing compiler gcc/7.3.0
  Starting job gcc-7.3.0-OpenMP-release
...
[ 97%] Linking CXX executable KokkosKernels_blas_openmp
CMakeFiles/KokkosKernels_blas_openmp.dir/openmp/Test_OpenMP_Blas3_trsm.cpp.o: In function `void KokkosBlas::trsm<Kokkos::View<double**, Kokkos::LayoutLeft, Kokkos::OpenMP>, Kokkos::View<double**, Kokkos::LayoutLeft, Kokkos::OpenMP> >(char const*, char const*, char const*, char const*, Kokkos::View<double**, Kokkos::LayoutLeft, Kokkos::OpenMP>::const_value_type&, Kokkos::View<double**, Kokkos::LayoutLeft, Kokkos::OpenMP> const&, Kokkos::View<double**, Kokkos::LayoutLeft, Kokkos::OpenMP> const&)':
Test_OpenMP_Blas3_trsm.cpp:(.text._ZN10KokkosBlas4trsmIN6Kokkos4ViewIPPdJNS1_10LayoutLeftENS1_6OpenMPEEEES7_EEvPKcS9_S9_S9_RNT0_16const_value_typeERKT_RKSA_[_ZN10KokkosBlas4trsmIN6Kokkos4ViewIPPdJNS1_10LayoutLeftENS1_6OpenMPEEEES7_EEvPKcS9_S9_S9_RNT0_16const_value_typeERKT_RKSA_]+0x55a): undefined reference to `KokkosBlas::Impl::TRSM<Kokkos::View<double const**, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::MemoryTraits<1u> >, Kokkos::View<double**, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::MemoryTraits<1u> >, false, true>::trsm(char const*, char const*, char const*, char const*, double const&, Kokkos::View<double const**, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::MemoryTraits<1u> > const&, Kokkos::View<double**, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::MemoryTraits<1u> > const&)'
collect2: error: ld returned 1 exit status

....
  Starting job gcc-7.3.0-Pthread-release
...
[ 97%] Linking CXX executable KokkosKernels_blas_threads
CMakeFiles/KokkosKernels_blas_threads.dir/threads/Test_Threads_Blas3_trsm.cpp.o: In function `void KokkosBlas::trsm<Kokkos::View<double**, Kokkos::LayoutLeft, Kokkos::Threads>, Kokkos::View<double**, Kokkos::LayoutLeft, Kokkos::Threads> >(char const*, char const*, char const*, char const*, Kokkos::View<double**, Kokkos::LayoutLeft, Kokkos::Threads>::const_value_type&, Kokkos::View<double**, Kokkos::LayoutLeft, Kokkos::Threads> const&, Kokkos::View<double**, Kokkos::LayoutLeft, Kokkos::Threads> const&)':
Test_Threads_Blas3_trsm.cpp:(.text._ZN10KokkosBlas4trsmIN6Kokkos4ViewIPPdJNS1_10LayoutLeftENS1_7ThreadsEEEES7_EEvPKcS9_S9_S9_RNT0_16const_value_typeERKT_RKSA_[_ZN10KokkosBlas4trsmIN6Kokkos4ViewIPPdJNS1_10LayoutLeftENS1_7ThreadsEEEES7_EEvPKcS9_S9_S9_RNT0_16const_value_typeERKT_RKSA_]+0x55a): undefined reference to `KokkosBlas::Impl::TRSM<Kokkos::View<double const**, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Threads, Kokkos::HostSpace>, Kokkos::MemoryTraits<1u> >, Kokkos::View<double**, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Threads, Kokkos::HostSpace>, Kokkos::MemoryTraits<1u> >, false, true>::trsm(char const*, char const*, char const*, char const*, double const&, Kokkos::View<double const**, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Threads, Kokkos::HostSpace>, Kokkos::MemoryTraits<1u> > const&, Kokkos::View<double**, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Threads, Kokkos::HostSpace>, Kokkos::MemoryTraits<1u> > const&)'
```

etc

**Comment by jjwilke** [comment]
Yeah, somehow I managed to rebase with no conflicts without actually adding `trsm` to the new ETI system.  Anyhow, ETI files are now generated for `trsm`, which should clear the error. The revision history got refactored to read a bit more sensibly - so you'll need a clean checkout.

On both develop and the branch, I'm getting `gemv` errors in the tests, though.

**Comment by ndellingwood** [comment]
Tests on kokkos-dev-2 look better:

```
#######################################################
PASSED TESTS
#######################################################
clang-8.0-Pthread_Serial-release build_time=48 run_time=133
cuda-10.1-Cuda_OpenMP-release build_time=151 run_time=123
gcc-7.3.0-OpenMP-release build_time=46 run_time=46
gcc-7.3.0-Pthread-release build_time=42 run_time=67
gcc-9.1-OpenMP-release build_time=51 run_time=46
gcc-9.1-Serial-release build_time=45 run_time=62
intel-18.0.5-OpenMP-release build_time=129 run_time=45
#######################################################
FAILED TESTS
#######################################################
clang-8.0-Cuda_OpenMP-release (build failed)
```

The failing clang build is due to `-Werror` triggered in the fallback gemm routine; talked to Christian about this, we're going to disable checking for the particular warning in cm_test_all_sandia, PR #558 has the change to script, which allows this test to pass.

**Comment by ndellingwood** [comment]
White spot-check testing with cm_test_all_sandia looks good.

gcc tests passed, I had to restart tests after with corrected arch options, below is output for xl and cuda:

```
${KOKKOSKERNELS_PATH}/scripts/cm_test_all_sandia ibm cuda --kokkos-path=${KOKKOS_PATH} --kokkoskernels-path=${KOKKOSKERNELS_PATH} --spot-check --num=1 --arch=Power8,Pascal60

Running on machine: white
Repository Status:  94ae52daca2f919cb4c9f61afbecaca7eca93a0b completely remove all ETI scripts and raw Makefiles


Going to test compilers:  ibm/16.1.0 cuda/9.2.88
Testing compiler ibm/16.1.0
  Starting job ibm-16.1.0-OpenMP-release
kokkos options:
kokkos devices: OpenMP
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Wsign-compare -Wtype-limits -Wuninitialized
  PASSED ibm-16.1.0-OpenMP-release
  Starting job ibm-16.1.0-Serial-release
kokkos options:
kokkos devices: Serial
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Wsign-compare -Wtype-limits -Wuninitialized
  PASSED ibm-16.1.0-Serial-release
Testing compiler cuda/9.2.88
  Starting job ibm-16.1.0-OpenMP_Serial-release
kokkos options:
kokkos devices: OpenMP,Serial
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Wsign-compare -Wtype-limits -Wuninitialized
  PASSED ibm-16.1.0-OpenMP_Serial-release
  Starting job cuda-9.2.88-Cuda_OpenMP-release
kokkos options:
kokkos devices: Cuda,OpenMP
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized
  PASSED cuda-9.2.88-Cuda_OpenMP-release

==============================================
Complete.
==============================================
  Starting job cuda-9.2.88-Cuda_Serial-release
kokkos options:
kokkos devices: Cuda,Serial
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized
  PASSED cuda-9.2.88-Cuda_Serial-release
```

**Comment by srajama1** [comment]
> Timing data on kokkos-dev-2, `make -j40`
> 
>     * Old ETI w/ TESTS
>       -real	2m38.381s
>       -user	42m4.556s
>       -sys	8m11.060s
> 
>     * Old ETI w/o TESTS
>       -real	1m56.462s
>       -user	28m27.473s
>       -sys	5m34.347s
> 
>     * New ETI w/ TESTS
>       -real	3m34.466s
>       -user	32m35.050s
>       -sys	3m58.608s
> 
>     * New ETI w/o TESTS
>       -real	2m49.847s
>       -user	18m18.393s
>       -sys	1m18.796s
> 
>     * New ETI w/ TESTS, one file per ETI instance
>       -real	1m56.433s
>       -user	44m59.249s
>       -sys	6m22.456s
> 
>     * New ETI w/o TESTS one file per ETI instance
>       -real	1m20.096s
>       -user	31m46.109s
>       -sys	3m44.812s

We pushed the one file per ETI instance change, right ? Thanks for keeping track of the times.

**Comment by jjwilke** [comment]
Yes, the most recent commit does one ETI instance per file for better parallelism.

**Comment by srajama1** [comment]
@jjwilke Thank you !

**Comment by srajama1** [comment]
@jjwilke Is this ready to go in to develop ?

**Comment by ndellingwood** [comment]
@srajama1 we need to convert testing to the cm_* scripts first, or else nearly all our nightlies will begin failing.

**Comment by jjwilke** [comment]
I believe this is ready. Obviously we'll want some extensive testing.

**Comment by ndellingwood** [comment]
@jjwilke the nightlies are now testing with the cmake build system, so that blocker is removed from this PR.
@srajama1 @iyamazaki when this PR is ready for merge the Makefile build system will be removed, so a cmake configure to build the supernodal sptrsv will be needed to exercise the capability.

**Comment by jjwilke** [comment]
I would merge the TPL PR #546, then merge this. I this branch has been rebased and should be ready to merge. Someone will want to make sure all the recent SPILUK, SPTRSV changes are still working in this branch.

**Comment by jjwilke** [comment]
@ndellingwood spot-check and spot-check-tpls are all passing on kokkos-dev-2. I think this is ready to go.

File: src/impl/generated_specializations_cpp/change [hunk 2]
```diff
@@ -0,0 +1,36 @@
+#! /usr/bin/env python
```
Comment (ndellingwood) [inline src/impl/generated_specializations_cpp/change]: I don't write python code often, wanted to check if this is python 2 vs 3 agnostic?

**Comment by ndellingwood** [comment]
@jjwilke arg, I dropped another merge conflict on you, sorry...

**Comment by ndellingwood** [comment]
@jjwilke if you have time to clean up the merge conflict (sorry for another) I think we're good to merge this in?

**Comment by jjwilke** [comment]
Rebased. Ready to go?

**Comment by ndellingwood** [comment]
Merging that was like our version of "the snap"

**Pull request merged**